src/instantx.cpp

```diff
@@ -878,17 +880,20 @@ bool CTxLockRequest::IsValid(bool fRequireUnspent) const
                 LogPrint("instantsend", "txLockRequest::IsValid -- Failed to find outpoint %s\n", txin.prevout.ToStringShort());
                 return false;
             }
-            LOCK(cs_main);
             BlockMap::iterator mi = mapBlockIndex.find(nHashOutpointConfirmed);
-            if(mi == mapBlockIndex.end()) {
-                // not on this chain?
+            if(mi == mapBlockIndex.end() || !mi->second) {
+                // shouldn't happen
                 LogPrint("instantsend", "txLockRequest::IsValid -- Failed to find block %s for outpoint %s\n", nHashOutpointConfirmed.ToString(), txin.prevout.ToStringShort());
                 return false;
             }
-            nPrevoutHeight = mi->second ? mi->second->nHeight : 0;
+            nPrevoutHeight = mi->second->nHeight;
+            nValue = txOutpointCreated.vout[txin.prevout.n].nValue;
```

Comment:
I would protect lines 890 and 893 with a check comparing vout.size with txin.prevout.n and returning false it it fails.

Maybe this can't happen but this depends on a lot of code (ie. various other functions) so imo it would be better to be safe and detect any problems here instead of crashing.